rfcs/20190821-nodejs-saved-model.md

```diff
@@ -0,0 +1,401 @@
+
+
+```
+TF_Session *session = TF_LoadSessionFromSavedModel(
+     session_options, run_options, export_dir, tags, tags_leng, graph,
+     metagraph, tf_status.status);
+```
+
+
+Pros:
+
+*   Using the available TF C API voids extra work to dive deep into the SavedModel structure and rebuild/execute the associated graph in JavaScript.
+*   New changes of SavedModel will be updated in the TF C library. The tfjs-node APIs will not be broken by SavedModel change.
+
+Cons:
+
+*   C API does not provide details and metadata about SavedModel
+
+
+##### Other option: parse saved_model.pb with protobuf in javascript
+
+Google’s Protocol Buffers is also available in javascript. It provides a [Protocol compiler](https://github.com/protocolbuffers/protobuf/releases) to translate the xxx.proto file to js file, and a JavaScript Protocol Buffers runtime library [google-protobuf](https://www.npmjs.com/package/google-protobuf) to construct and parse the messages.
+
+To use Protocol Buffers in javascript, first the saved_model.proto file need to be translated:
+
+
+```
+$ protoc --js_out=import_style=commonjs,binary:. saved_model.proto
+```
+
+
+The above command will translate the [saved_model.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/saved_model.proto) file to saved_model_pb.js file. However the saved_model.proto file depends on [meta_graph.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/meta_graph.proto). The meta_graph.proto file depends on [graph.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/graph.proto), [op_def.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op_def.proto), [tensor_shape.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor_shape.proto), [types.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/types.proto), [saved_object_graph.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/saved_object_graph.proto), [savef.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/saver.proto), and these proto files depend on more proto files in the tensorflow repo.
+
+Thus almost all the proto files in tensorflow are required to be translated to js version. This can be achieved by clustering all the proto files in [tensorflow repo](https://github.com/tensorflow/tensorflow) together into one folder,  running the command above and moving the generated translated_pb.js files into tfjs-node repo.
```

Comment:
Why is this different from just using bazel to generate js proto rules?

Reply:
Using the [Protocol Compiler](https://github.com/bazelbuild/bazel/tree/master/third_party/protobuf/3.6.1/js#protocol-buffers---googles-data-interchange-format) should be the same as using bazel to generate js proto rules. It will only be used once to generate the saved_model_proto.js file and will not be used by users.